File: examples/java/operations/map/nested-map.java
import java.util.List;
import software.amazon.lambda.durable.DurableContext;
import software.amazon.lambda.durable.DurableHandler;
import software.amazon.lambda.durable.TypeToken;
import software.amazon.lambda.durable.model.MapResult;

public class NestedMap extends DurableHandler<List<Region>, List<List<String>>> {
    record Region(String name, List<String> items) {}

    @Override
    public List<List<String>> handleRequest(List<Region> regions, DurableContext context) {
        MapResult<List<String>> result = context.map(
                "process-regions",
                regions,
                new TypeToken<List<String>>() {},
                (region, index, ctx) -> {
                    MapResult<String> inner = ctx.map(
                            "process-" + region.name(),
                            region.items(),
                            String.class,
                            (item, i, innerCtx) -> innerCtx.step(
                                    "item-" + i, String.class, s -> item.toUpperCase()));
                    return inner.succeeded();
                });

        return result.succeeded();
    }
}

File: examples/csharp/operations/map/nested-map.cs
using Amazon.Lambda.Core;
using Amazon.Lambda.DurableExecution;

public class NestedMapExample
{
    public Task<DurableExecutionInvocationOutput> Handler(
        DurableExecutionInvocationInput input, ILambdaContext context)
        => DurableFunction.WrapAsync<RegionEvent, IReadOnlyList<IReadOnlyList<string>>>(
            Workflow, input, context);

    private async Task<IReadOnlyList<IReadOnlyList<string>>> Workflow(
        RegionEvent input, IDurableContext ctx)
    {
        IBatchResult<IReadOnlyList<string>> result = await ctx.MapAsync(
            input.Regions,
            async (regionCtx, region, index, regions, ct) =>
            {
                IBatchResult<string> inner = await regionCtx.MapAsync(
                    region.Items,
                    async (itemCtx, item, i, items, innerCt) =>
                        await itemCtx.StepAsync(
                            async (_, _) => item.ToUpperInvariant(),
                            name: $"item-{i}"),
                    name: $"process-{region.Name}");
                return inner.GetResults();
            },
            name: "process-regions");

        return result.GetResults();
    }
}

public record Region(string Name, IReadOnlyList<string> Items);
public record RegionEvent(IReadOnlyList<Region> Regions);
